include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -401,12 +426,160 @@ namespace PETScWrappers
                  PMatrixType,
                  Mat>)&&(concepts::is_dealii_petsc_matrix_type<AMatrixType> ||
                          std::constructible_from<AMatrixType, Mat>))
-  unsigned int TimeStepper<VectorType, PMatrixType, AMatrixType>::solve(
-    VectorType &y)
+  void TimeStepper<VectorType, PMatrixType, AMatrixType>::setup_callbacks()
   {
-    const auto ts_prestage = [](TS ts, PetscReal) -> PetscErrorCode {
+    const auto ts_resize_setup =
+      [](TS ts, PetscInt it, PetscReal t, Vec x, PetscBool *flg, void *ctx)
+      -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      auto user = static_cast<TimeStepper *>(ctx);
+
+      VectorType xdealii(x);
+      bool       resize = false;
+
+      const int lineno = __LINE__;
+      const int err    = call_and_possibly_capture_ts_exception(
+        user->prepare_for_coarsening_and_refinement,
+        user->pending_exception,
+        []() -> void {},
```

Comment:
Is this actually what you want? This is the `recoverable_action` argument, and it is used as follows:
```
    try
      {
        f(std::forward<Args>(args)...);
        eptr = nullptr;
        return 0;
      }
    // In case of a recoverable exception call the action if present:
    catch (const RecoverableUserCallbackError &)
      {
        if (recoverable_action)
          {
            // recover and eat exception
            recoverable_action();
            eptr = nullptr;
            return 0;
          }
        else
          {
            // No action provided.
            // capture exception and return -1
            eptr = std::current_exception();
            return -1;
          }
      }
```
So by passing an empty function here, you are saying "yes, I know how to deal with this error", but then you don't actually do anything with it.

I *think* what you want to do is not pass a function at all, so that you get into the `else` branch:
```suggestion
        {},
```

Reply:
done here and in a few other places. Thanks